mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt

```diff
@@ -271,59 +274,78 @@ class ModelSyncGuiFactory : ToolWindowFactory, Disposable {
-                    if (branchModel.size > 0) {
-                        branchModel.selectedItem = branchModel.getElementAt(0)
-                        populateModuleCB()
-                    }
+                val client = existingConnectionsModel.selectedItem as ModelClientV2
+                val repositoryId = repoModel.selectedItem as RepositoryId
+                val branches = client.listBranches(repositoryId)
+
+                branchModel.removeAllElements()
+                branchModel.addAll(branches)
+                if (branchModel.size > 0) {
+                    branchModel.selectedItem = branchModel.getElementAt(0)
+                    populateModuleCB()
                 }
             }
         }
 
-        private fun populateModuleCB() {
+        private suspend fun populateModuleCB() {
             if (existingConnectionsModel.size != 0 && repoModel.size != 0 && branchModel.size != 0) {
-                CoroutineScope(Dispatchers.Default).launch {
-                    val branch =
-                        (existingConnectionsModel.selectedItem as ModelClientV2).getReplicatedModel(branchModel.selectedItem as BranchReference)
-                            .start()
-                    branch.runRead {
-                        moduleModel.removeAllElements()
-                        val children = branch.getRootNode().allChildren.map {
-                            val name = it.getPropertyValue(BuiltinLanguages.jetbrains_mps_lang_core.INamedConcept.name)
-                                ?: it.toString()
-                            INodeWithName(it, name)
-                        }
-                        moduleModel.addAll(children.toList())
-                    }
-                    if (moduleModel.size > 0) {
-                        moduleModel.selectedItem = moduleModel.getElementAt(0)
+                val client = existingConnectionsModel.selectedItem as ModelClientV2
+                val branchReference = branchModel.selectedItem as BranchReference
+
+                val moduleNodes = client.query(branchReference) {
+                    it.allChildren().ofConcept(BuiltinLanguages.MPSRepositoryConcepts.Module).toList()
+                }.map {
+                    val name = it.getPropertyValue(BuiltinLanguages.jetbrains_mps_lang_core.INamedConcept.name)
+                        ?: it.toString()
+                    val id = it.getPropertyValue(BuiltinLanguages.MPSRepositoryConcepts.Module.id) ?: ""
+                    ModuleIdWithName(id, name)
+                }
+
+                moduleModel.removeAllElements()
+                moduleModel.addAll(moduleNodes.toList())
+                if (moduleModel.size > 0) {
+                    moduleModel.selectedItem = moduleModel.getElementAt(0)
+                }
+            }
+        }
+
+        @Synchronized
+        private fun callOnlyIfNotFetching(action: suspend () -> Unit) {
+            if (!isFetchingModulesListFromServer.get()) {
```

Comment:
Should the combo boxes be disabled for user input while fetching? And should and indicator be shown in the UI?
Maybe this is out of scope for this PR. If so, could you add a short note in the code about the planned long-term behaviour?

Reply:
Fixed in 71d44f0bf1c3784025660898d612f589af273833.

On the long term we have more UI feedback on the roadmap, see [MODELIX-621](https://issues.modelix.org/issue/MODELIX-621) and [MODELIX-725](https://issues.modelix.org/issue/MODELIX-725).